cpp/src/arrow/scalar_test.cc

```diff
@@ -1984,4 +1985,41 @@ TEST_F(TestExtensionScalar, ValidateErrors) {
   AssertValidationFails(scalar);
 }
 
+template <typename T>
+class TestScalarScratchSpace : public ::testing::Test {
+ public:
+  TestScalarScratchSpace() = default;
+};
+
+TYPED_TEST_SUITE(TestScalarScratchSpace, BaseBinaryOrBinaryViewLikeArrowTypes);
+
+// GH-40069: race condition when filling the scratch space of a scalar in parallel.
```

Comment:
Does this test reproduce the thread sanitizer failure on the thread sanitizer CI job?

Reply:
I didn't see a TSAN CI job in current job list. Either I'm not sure how to trigger one.

But I confirm that the `ctest` command failed in my local, built with `-DARROW_USE_TSAN=ON -DARROW_JEMALLOC=OFF -DARROW_MIMALLOC=OFF`.

What's interesting, though I'm not exactly sure why, is that gtest actually reported PASSED because the TSAN error is just WARNING printed to stderr. But the test program did aborted (returned non-zero code) and ctest considered it failed.

Reply:
Oh BTW, if we want to see a TSAN CI job fail, I'll have to revert the change in `data.cc`, which contains the fix already.